filler.go

```diff
@@ -0,0 +1,116 @@
+package rocket
+
+import (
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"reflect"
+)
+
+type (
+	contextFiller interface {
+		fill(context reflect.Value) error
+	}
+	nextFiller struct {
+		next contextFiller
+	}
+	routeFiller struct {
+		nextFiller
+		h  *handler
+		rs []string
+	}
+	queryFiller struct {
+		nextFiller
+		h     *handler
+		query url.Values
+	}
+	jsonFiller struct {
+		nextFiller
+		h   *handler
+		req *http.Request
+	}
+	formFiller struct {
+		nextFiller
+		h    *handler
+		form url.Values
+	}
+)
+
+func (n *nextFiller) nextFill(ctx reflect.Value) error {
+	if n.next != nil {
+		return n.next.fill(ctx)
+	}
+	return nil
+}
+func (r *routeFiller) fill(ctx reflect.Value) error {
+	for idx, route := range r.h.routes {
+		if isParameter(route) {
+			param := r.rs[len(r.rs)-len(r.h.routes)+idx]
+			index := r.h.routeParams[idx]
+			v := parseParameter(ctx.Elem().Field(index), param)
+			ctx.Elem().Field(index).
+				Set(v)
+		}
+	}
+	if r.h.matchedPathIndex != -1 {
+		i := r.h.routeParams[r.h.matchedPathIndex]
+		ctx.Elem().Field(i).
+			Set(reflect.ValueOf(r.h.matchedPath))
+	}
+	return r.nextFill(ctx)
+}
+func (q *queryFiller) fill(ctx reflect.Value) error {
+	for k, idx := range q.h.queryParams {
+		field := ctx.Elem().Field(idx)
+		if v, ok := q.query[k]; ok {
+			p := v[0]
+			value := parseParameter(field, p)
+			field.Set(value)
+		}
+	}
+	return q.nextFill(ctx)
+}
+func (j *jsonFiller) fill(ctx reflect.Value) error {
+	if j.h.expectJsonRequest {
+		v := ctx.Interface()
+		err := json.NewDecoder(j.req.Body).Decode(v)
+		if err != nil {
+			return err
+		}
+		ctx.Elem().Set(reflect.ValueOf(v).Elem())
+		return nil
+	}
+	return j.nextFill(ctx)
+}
+func (f *formFiller) fill(ctx reflect.Value) error {
+	for k, idx := range f.h.formParams {
+		if v, ok := f.form[k]; ok {
+			field := ctx.Elem().Field(idx)
+			p := v[0]
+			value := parseParameter(field, p)
+			field.Set(value)
+		}
+	}
+	return f.nextFill(ctx)
+}
+
```

Comment:
Better to add default filler chain, users might break the assuming, route -> query -> json -> form is required!